### PR TITLE
Add initial stack drift command

### PR DESCRIPTION
**Pull request by alex-harvey-z3q**
Add a `detect-stack-drift` command and tests.

The `detect-stack-drift` command calls the `detect_stack_drift` Boto3 call, takes the Detector Id returned, then waits for `describe_stack_drift_detection_status(StackDriftDetectionId=detector_id)` to complete, and then finally returns `describe_stack_resource_drifts` as a JSON document.

The output is formatted such that it can be passed directly into a tool like `jq`. 

If `--debug` is passed, sceptre also will provide feedback on the detection progress.

### CODE REVIEW

**Comment by alex-harvey-z3q** [comment]
Example output:

### debug mode

```
[aws:foxsports-gitops-test] /src $ sceptre --debug --dir=network --var-file=/env/common-env.yaml --var-file=/env/network.yaml detect-stack-drift network.yaml
...
[2021-08-06 16:21:25] - Boto3 session created                                                                                                                                     
[2021-08-06 16:21:25] - vpc - Waiting for drift detection                                                                                                                         [2021-08-06 16:21:25] - vpc - Detecting Stack Drift Detection Status                                                                                                              
[2021-08-06 16:21:25] - vpc - StackDriftDetectionId - 86c17670-f67e-11eb-a5fc-067ab790b156
[2021-08-06 16:21:25] - vpc - DetectionStatus - DETECTION_IN_PROGRESS                    
[2021-08-06 16:21:25] - vpc - DetectionStatusReason - User Initiated
[2021-08-06 16:21:35] - vpc - Waiting for drift detection 
[2021-08-06 16:21:35] - vpc - Detecting Stack Drift Detection Status                                                                                                              
[2021-08-06 16:21:35] - vpc - StackDriftDetectionId - 86c17670-f67e-11eb-a5fc-067ab790b156
[2021-08-06 16:21:35] - vpc - DetectionStatus - DETECTION_COMPLETE                       
[2021-08-06 16:21:35] - vpc - StackDriftStatus - IN_SYNC
[2021-08-06 16:21:35] - vpc - Detecting Stack Drift                                                                                                                               
[2021-08-06 16:21:35] - nacl - Detecting Stack Drift                                                                                                                              [2021-08-06 16:21:35] - nacl - Waiting for drift detection                                                                                                                        [2021-08-06 16:21:35] - nacl - Detecting Stack Drift Detection Status                                                                                                             
[2021-08-06 16:21:35] - nacl - StackDriftDetectionId - 8cea4e50-f67e-11eb-8a7c-061cb8f1bfb8                                                                                       [2021-08-06 16:21:35] - nacl - DetectionStatus - DETECTION_IN_PROGRESS                                                                                                            [2021-08-06 16:21:35] - nacl - DetectionStatusReason - User Initiated                                                                                                             
[2021-08-06 16:21:45] - nacl - Waiting for drift detection
[2021-08-06 16:21:45] - nacl - Detecting Stack Drift Detection Status                    
[2021-08-06 16:21:45] - nacl - StackDriftDetectionId - 8cea4e50-f67e-11eb-8a7c-061cb8f1bfb8
[2021-08-06 16:21:45] - nacl - DetectionStatus - DETECTION_COMPLETE                                                                                                               
[2021-08-06 16:21:45] - nacl - StackDriftStatus - IN_SYNC                                                                                                                         
[2021-08-06 16:21:45] - nacl - Detecting Stack Drift   
[2021-08-06 16:21:45] - network - Detecting Stack Drift                                  
[2021-08-06 16:21:45] - network - Waiting for drift detection
[2021-08-06 16:21:45] - network - Detecting Stack Drift Detection Status
[2021-08-06 16:21:45] - network - StackDriftDetectionId - 931a7930-f67e-11eb-a48f-0a44c1f6bcd4                                                                                    [2021-08-06 16:21:45] - network - DetectionStatus - DETECTION_IN_PROGRESS                                                                                                         [2021-08-06 16:21:45] - network - DetectionStatusReason - User Initiated                                                                                                          
[2021-08-06 16:21:56] - network - Waiting for drift detection                                                                                                                     [2021-08-06 16:21:56] - network - Detecting Stack Drift Detection Status                                                                                                          [2021-08-06 16:21:56] - network - StackDriftDetectionId - 931a7930-f67e-11eb-a48f-0a44c1f6bcd4                                                                                    
[2021-08-06 16:21:56] - network - DetectionStatus - DETECTION_IN_PROGRESS
[2021-08-06 16:21:56] - network - DetectionStatusReason - User Initiated
[2021-08-06 16:22:06] - network - Waiting for drift detection
[2021-08-06 16:22:06] - network - Detecting Stack Drift Detection Status
[2021-08-06 16:22:06] - network - StackDriftDetectionId - 931a7930-f67e-11eb-a48f-0a44c1f6bcd4                                                                                    
[2021-08-06 16:22:06] - network - DetectionStatus - DETECTION_FAILED                     
[2021-08-06 16:22:06] - network - DetectionStatusReason - {"Summary":"Failed to detect drift on resources [PrivateSubnetb]","Failures":[{"Resource":"PrivateSubnetb","FailureReaso
n":"Rate exceeded"}]}   
[2021-08-06 16:22:06] - network - StackDriftStatus - DRIFTED
{                                                                                                                                                                                 
  "gitops-test-vpc": {                                                                                                                                                            
    "ResponseMetadata": {                                                                                                                                                         
      "HTTPHeaders": {                                                                                                                                                            
        "content-length": "2491",                                                                                                                                                 
        "content-type": "text/xml",                                                                                                                                               
        "date": "Fri, 06 Aug 2021 04:43:28 GMT",                                                                                                                                  
        "vary": "accept-encoding",                                                                                                                                                
        "x-amzn-requestid": "e18f763f-20cc-4ec1-8f95-37de41d3aeac"                                                                                                                
      },                                                                                                                                                                          
      "HTTPStatusCode": 200,                                                                                                                                                      
      "RequestId": "e18f763f-20cc-4ec1-8f95-37de41d3aeac",                                                                                                                        
      "RetryAttempts": 0                                                                                                                                                          
    },                                                                                                                                                                            
    "StackResourceDrifts": [                                                                                                                                                      
      {                                                                                                                                                                           
        "ActualProperties": "{\"CidrBlock\":\"10.90.0.0/19\",\"EnableDnsHostnames\":true,\"EnableDnsSupport\":true,\"Tags\":[{\"Key\":\"Department\",\"Value\":\"technology\"},{\"
Key\":\"Creator\",\"Value\":\"sceptre\"},{\"Key\":\"Contact\",\"Value\":\"devops@foxsports.com.au\"},{\"Key\":\"Name\",\"Value\":\"gitops-test\"},{\"Key\":\"Version\",\"Value\":\
"1.0.0\"},{\"Key\":\"Environment\",\"Value\":\"test\"},{\"Key\":\"Project\",\"Value\":\"GitOps\"},{\"Key\":\"Team\",\"Value\":\"DevOps\"}]}",                                     
        "ExpectedProperties": "{\"CidrBlock\":\"10.90.0.0/19\",\"EnableDnsHostnames\":true,\"EnableDnsSupport\":true,\"Tags\":[{\"Key\":\"Department\",\"Value\":\"technology\"},{
\"Key\":\"Creator\",\"Value\":\"sceptre\"},{\"Key\":\"Contact\",\"Value\":\"devops@foxsports.com.au\"},{\"Key\":\"Name\",\"Value\":\"gitops-test\"},{\"Key\":\"Version\",\"Value\"
:\"1.0.0\"},{\"Key\":\"Environment\",\"Value\":\"test\"},{\"Key\":\"Project\",\"Value\":\"GitOps\"},{\"Key\":\"Team\",\"Value\":\"DevOps\"}]}",                                   
        "LogicalResourceId": "VPC",                                                                                                              
etc
```

### pipe to jq

Screen Shot 2021-08-06 at 4 14 22 pm

Screen Shot 2021-08-06 at 4 12 55 pm

File: sceptre/plan/actions.py [hunk 1]
```diff
@@ -614,18 +614,127 @@ def diff(self, differ="difflib"):
...
+        :param detect_id: The drift detection ID.
+        :type detect_id: str
+
+        :returns: The drift status.
+        :rtype: str
+        """
+        timeout = 300
+        elapsed = 0
+
+        def print_status():
```
Comment (mrowlingfox) [inline sceptre/plan/actions.py]: Not a huge fan of the nested functions. Either refactor it out or just keep it inline

File: pyproject.toml [hunk 2]
```diff
@@ -19,7 +19,7 @@ exclude_lines = [
   "if TYPE_CHECKING:",
 ]
 ignore_errors = true
-fail_under = 89.0
+fail_under = 87.0
```
Comment (mrowlingfox) [inline pyproject.toml]: Has coverage still dipped this low after you added unit tests?

Comment (alex-harvey-z3q) [reply pyproject.toml]: Actually the merge-keys PR accidentally removed some tests in `test_cli.py`. That's fixed now.

File: sceptre/plan/actions.py (hunk 1 above)
Comment (alex-harvey-z3q) [reply sceptre/plan/actions.py]: done

File: sceptre/plan/actions.py [hunk 3]
```diff
@@ -614,18 +614,130 @@ def diff(self, differ="difflib"):
...
-            return_val = self._diff_by_dictdiffer(
+            response = self._diff_by_dictdiffer(
                 remote_template_stream,
                 local_template_stream
             )
 
-        return [self.stack.external_name, return_val]
+        return [self.stack.external_name, response]
+
+    @add_stack_hooks
```
Comment (mrowlingfox) [inline sceptre/plan/actions.py]: Can you think of a usecase where we need hooks?

File: pyproject.toml (hunk 2 above)
Comment (mrowlingfox) [reply pyproject.toml]: Let's bump this up again then

**Review by mrowlingfox — COMMENTED** [review]
Looks ok, but I haven't actually tested yet

File: sceptre/plan/actions.py (hunk 3 above)
Comment (alex-harvey-z3q) [reply sceptre/plan/actions.py]: Nope, removed.

**Review by mrowlingfox — APPROVED** [review]

**Pull request merged**